### PR TITLE
Kafka 4060 remove zk client dependency in kafka streams

### PR DESCRIPTION
**Pull request by hjafarpour**
Removed Zookeeper client from the Kafka Streams. The internal topic manager now will use Kafka Client to create/delete internal topics.

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java [hunk 1]
```diff
@@ -133,112 +129,179 @@
 
 
 
+    public static final String SECURITY_PROTOCOL_CONFIG = CommonClientConfigs.SECURITY_PROTOCOL_CONFIG;
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:143]: I don't think we should add these configs to the public interface. Developers can already add Consumer & Producer Configs to the StreamsConfig - they both have all of the configs defined here and we should just use those. For example have a look at `StreamsConfig.getProducerConfigs(...)`

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java [hunk 2]
```diff
@@ -133,112 +129,179 @@
...
-                                        Type.STRING,
-                                        "",
-                                        Importance.LOW,
-                                        APPLICATION_SERVER_DOC)
-                                .define(ROCKSDB_CONFIG_SETTER_CLASS_CONFIG,
-                                        Type.CLASS,
-                                        null,
-                                        Importance.LOW,
-                                        ROCKSDB_CONFIG_SETTER_CLASS_DOC);
+                Type.STRING,
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:170]: nit: formatting

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultKafkaClientSupplier.java [hunk 3]
```diff
@@ -42,4 +43,14 @@
...
         return new KafkaConsumer<>(config, new ByteArrayDeserializer(), new ByteArrayDeserializer());
     }
+
+    /**
+     * Creates a default StreamsKafkaClient. For now there is no need to expose this in the KafkaClientSupplier interface.
+     *
+     * @param config
+     * @return
+     */
+    public StreamsKafkaClient getStreamKafkaClient(StreamsConfig config) {
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultKafkaClientSupplier.java]: I don't think this is necessary to add here. AFAICT `StreamsKafkaClient` is only used in `InternalTopicManager` and it can just be constructed there

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java [hunk 4]
```diff
@@ -17,242 +17,225 @@
...
+        this.kafkaClient = null;
         this.replicationFactor = 0;
     }
 
-    public InternalTopicManager(String zkConnect, int replicationFactor) {
-        this.zkClient = new ZkClient(zkConnect, 30 * 1000, 30 * 1000, new ZKStringSerializer());
-        this.replicationFactor = replicationFactor;
+    public InternalTopicManager(StreamsConfig config) {
+        this.config = config;
+        this.streamsKafkaClient = new DefaultKafkaClientSupplier().getStreamKafkaClient(config);
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java]: `new StreamsKafkaClient(config)` ?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java [hunk 5]
```diff
@@ -17,242 +17,225 @@
...
+        this.kafkaClient = streamsKafkaClient.getKafkaClient();
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java]: Rather than getting the `kafkaClient` why not push the functionality down to `StreamsKafkaClient` ? Hide the details and make it useful such that the internals aren't exposed?

**Comment by dguy** [comment]
@hjafarpour i haven't had the opportunity to go over the diff properly, but the test failure in the build is genuine so it needs to be fixed. I also noticed a few `// TODO` items that should be fixed/removed

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java [hunk 6]
```diff
@@ -17,242 +17,225 @@
...
-            });
+    /**
+     * Check if the topic exists already
+     *
+     * @param topic
+     * @param brokerNode
+     * @return
+     */
+    private boolean topicExists(String topic, Node brokerNode) {
+        MetadataRequest metadataRequest = new MetadataRequest(Arrays.asList(topic));
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java]: I'm wondering if a lot of this code should be in `StreamsKafkaClient` or perhaps remove `StreamsKafkaClient` completely and construct the underlying `NetworkClient` in `InternalTopicManager`. If the functionality is something we can/will use elsewhere then I'd put it in `StreamsKafkaClient`

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java [hunk 7]
```diff
@@ -17,242 +17,225 @@
...
 
-            log.debug("Read partitions {} for topic {} from ZK in partition assignor.", partitions, topic);
+            }
+        };
+
+        // Send the async request to check the topic existance
+        ClientRequest clientRequest = new ClientRequest(new SystemTime().milliseconds(), true, send, callback);
+        SystemTime systemTime = new SystemTime();
+        int iterationCount = 0;
+        while (iterationCount < 5) { // TODO: Set this later
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java]: I'd probably do something like:

```
while (!kafka.ready(...)  && iterationCount++ < MAX_ITERATIONS) {
   kafkaClient.poll(...)
}
if (kafkaClient.ready(...) {
    kafkaClient.send(...)
}
```

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java [hunk 8]
```diff
@@ -17,242 +17,225 @@
...
+            public void onComplete(ClientResponse response) {
 
-            log.debug("Read partitions {} for topic {} from ZK in partition assignor.", partitions, topic);
+            }
+        };
+
+        // Send the async request to check the topic existance
+        ClientRequest clientRequest = new ClientRequest(new SystemTime().milliseconds(), true, send, callback);
+        SystemTime systemTime = new SystemTime();
+        int iterationCount = 0;
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java]: iterationCount is never incremented so the loop below is essentially `while(true)`

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java [hunk 9]
```diff
@@ -17,242 +17,225 @@
...
+            topicDetails = new CreateTopicsRequest.TopicDetails(numPartitions, (short) replicationFactor);
         }
 
-        // try to write to ZK with open ACL
-        try {
-            Map<String, Object> dataMap = new HashMap<>();
-            dataMap.put("version", 1);
-            dataMap.put("partitions", assignment);
-            String data = mapper.writeValueAsString(dataMap);
+        Map<String, CreateTopicsRequest.TopicDetails> topics = new HashMap<>();
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java]: `topicCreateRequests` ?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java [hunk 10]
```diff
@@ -17,242 +17,225 @@
...
+        RequestCompletionHandler callback = new RequestCompletionHandler() {
+            public void onComplete(ClientResponse response) {
+                handleResponse(response);
+            }
+        };
+
+        ClientRequest clientRequest = new ClientRequest(new SystemTime().milliseconds(), true, send, callback);
+        SystemTime systemTime = new SystemTime();
+        int iterationCount = 0;
+        while (iterationCount < 5) { // TODO: Set this later
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java]: Same as above comments on the loop. `iterationCount` is never incremented and i think the loop could be simplified a bit. Also, as it is the same logic it is probably worth extracting it into a method, i.e, `isClientReady` or sth and removing the duplication

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java [hunk 11]
```diff
@@ -451,19 +459,19 @@ public Subscription subscription(Set<String> topics) {
 
 
                 assignmentSuppliers.add(new AssignmentSupplier(consumer,
-                                                               active,
-                                                               standby,
-                                                               endPointMap,
-                                                               activePartitions));
+                        active,
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java:459]: nit: formatting

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java [hunk 12]
```diff
@@ -493,8 +501,8 @@ public Subscription subscription(Set<String> topics) {
 
         Assignment get() {
             return new Assignment(activePartitions, new AssignmentInfo(active,
-                                                                       standby,
-                                                                       endPointMap).encode());
+                    standby,
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java:501]: nit: formatting

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java [hunk 13]
```diff
@@ -524,7 +532,7 @@ public void onAssignment(Assignment assignment) {
             } else {
                 TaskAssignmentException ex = new TaskAssignmentException(
                         "failed to find a task id for the partition=" + partition.toString() +
-                        ", partitions=" + partitions.size() + ", assignmentInfo=" + info.toString()
+                                ", partitions=" + partitions.size() + ", assignmentInfo=" + info.toString()
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java]: nit: formatting

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java [hunk 14]
```diff
@@ -589,7 +597,7 @@ private void ensureCopartitioning(Set<String> copartitionGroup, Set<String> inte
         for (String topic : internalTopics) {
             if (copartitionGroup.contains(topic)) {
                 internalSourceTopicToTaskIds
-                    .put(topic, Collections.singleton(new TaskId(-1, numPartitions)));
+                        .put(topic, Collections.singleton(new TaskId(-1, numPartitions)));
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java:618]: nit: formatting

File: streams/src/test/java/org/apache/kafka/streams/integration/InternalTopicIntegrationTest.java [hunk 15]
```diff
@@ -69,42 +70,57 @@ public static void startKafkaCluster() throws Exception {
...
     /**
-     * Validates that any state changelog topics are compacted
-     * @return true if topics have a valid config, false otherwise
+     * Check if the given topic exists.
+     *
+     * @param topic
+     * @param config
+     * @return
      */
-    private boolean isUsingCompactionForStateChangelogTopics() {
```
Comment (enothereska) [inline streams/src/test/java/org/apache/kafka/streams/integration/InternalTopicIntegrationTest.java]: Are we still checking if the topic is compacted with the new code?

Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/integration/InternalTopicIntegrationTest.java]: I think we should, and with #1792 maybe this test needs to be updated a bit again, but still by default (user does not overwrite anything) the changelog should be compacted.

**Comment by guozhangwang** [comment]
Could you also change the title to `KAFKA-4060: ...` since otherwise the merge tool would not be able to locate the corresponding JIRA ticket while merging.

**Comment by guozhangwang** [comment]
@hjafarpour Could you try to address the comments?

**Comment by hjafarpour** [comment]
@guozhangwang @dguy @enothereska Just saw the comments! Seems that github notifications were not sent to my email. Will look into the comments and address them soon!

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultKafkaClientSupplier.java (hunk 3 above)
Comment (hjafarpour) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultKafkaClientSupplier.java]: Removing it from DefaultKafkaClientSupplier and instantiate it directly as you suggested.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java (hunk 4 above)
Comment (hjafarpour) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java]: Changed as you suggested.

**Comment by guozhangwang** [comment]
Also ping @mjsax to review.

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java (hunk 1 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:143]: Just had another discussion with @hjafarpour offline. And after that I am now lining towards keeping these configs on the `StreamsConfig` layer as well for the following two reasons:
1. All these `CommonClientConfigs` are defined both in producer and consumer configs as well, but their config definitions may have different default values. It is not clear which default values we should use then if users do not specify customized values.
2. Right now we have the `prefix` for producer and consumer to differentiate those common configs. So following this pattern if we define the same config `X` also in StreamsConfig, we can then allow users to specify different values for `X` for streams' admin client, and producer / consumer client separately, as:

```
producer.X: 100
consumer.X: 200
X: 300
```

And without the prefixed overriding:

```
X: 300
```

Then all three clients will use the same value, which is reasonable.

Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:143]: Makes sense, but should this be a different PR than?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsKafkaClient.java [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+
+import java.net.InetSocketAddress;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Created by hojjat on 8/29/16.
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsKafkaClient.java]: nit: remove

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsKafkaClient.java [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+    private static final Logger log = LoggerFactory.getLogger(StreamsKafkaClient.class);
+    private Metadata metadata;
+    private Metrics metrics;
+    private Selector selector;
+    private ChannelBuilder channelBuilder;
+    private KafkaClient kafkaClient;
+
+    public StreamsKafkaClient(StreamsConfig config) {
+
+        Time time = new SystemTime();
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsKafkaClient.java:85]: For testing, we might want to be able to use `MockTime` here. Not sure if this would be easy with this (not sure how to change it either -- to be honest)

File: streams/src/test/java/org/apache/kafka/streams/integration/InternalTopicIntegrationTest.java [hunk 18]
```diff
@@ -69,42 +70,57 @@ public static void startKafkaCluster() throws Exception {
...
+            public void onComplete(ClientResponse response) {
+
+            }
+        };
+
+        // Send the async request to check the topic existance
+        ClientRequest clientRequest = new ClientRequest(new SystemTime().milliseconds(), true, send, callback);
+        SystemTime systemTime = new SystemTime();
+        int iterationCount = 0;
+        while (iterationCount < 5) { // TODO: Set this later
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/InternalTopicIntegrationTest.java]: This might be instable in Jenkins.

File: streams/src/test/java/org/apache/kafka/streams/integration/InternalTopicIntegrationTest.java [hunk 19]
```diff
@@ -69,42 +70,57 @@ public static void startKafkaCluster() throws Exception {
...
+                break;
+            } else {
+                // If the client is not ready call poll to make the client ready
+                kafkaClient.poll(config.getLong(StreamsConfig.POLL_MS_CONFIG), new SystemTime().milliseconds());
+            }
+        }
+
+        // Process the response
+        iterationCount = 0;
+        while (iterationCount < 5) { // TODO: Set this later
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/InternalTopicIntegrationTest.java]: As above.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsKafkaClient.java [hunk 20]
```diff
@@ -0,0 +1,105 @@
...
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Created by hojjat on 8/29/16.
+ */
+public class StreamsKafkaClient {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsKafkaClient.java:66]: Why do we need this? Seems to be a wrapper for `NetworkClient`, but does not add too much value IMHO.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java [hunk 21]
```diff
@@ -17,242 +17,225 @@
...
-        String data = zkClient.readData(ZK_TOPIC_PATH + "/" + topic, true);
-
-        if (data == null) return null;
+        Node brokerNode = kafkaClient.leastLoadedNode(new SystemTime().milliseconds());
 
-        try {
-            ObjectMapper mapper = new ObjectMapper();
+        // Delete the topic if it exists before.
+        boolean topicExists = topicExists(topic, brokerNode);
+        if (topicExists) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java]: Why did you remove the check for number of partitions?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java (hunk 7 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java]: I guess we should throw or log, too, if `MAX_ITERATIONS` gets hit.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java [hunk 22]
```diff
@@ -17,242 +17,225 @@
...
+            deleteTopic(topic, brokerNode);
+            log.debug("Deleted: " + topic);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java]: Deleting is async, right? I guess we should make sure it got deleted, otherwise, `createTopic` might fail.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java [hunk 23]
```diff
@@ -17,242 +17,225 @@
...
+        }
+        // Create a new topic
+        createTopic(topic, numPartitions, replicationFactor, compactTopic, brokerNode);
+        log.debug("Created: " + topic);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java]: Should be check, if the topic is available?

File: build.gradle [hunk 24]
```diff
@@ -495,7 +495,7 @@ project(':core') {
 
   task siteDocsTar(dependsOn: ['genProtocolErrorDocs', 'genProtocolApiKeyDocs', 'genProtocolMessageDocs',
                                'genProducerConfigDocs', 'genConsumerConfigDocs', 'genKafkaConfigDocs',
-                               'genTopicConfigDocs', ':connect:runtime:genConnectConfigDocs', 
+                               'genTopicConfigDocs', ':connect:runtime:genConnectConfigDocs',
```
Comment (dguy) [inline build.gradle:498]: un-intenional?

File: gradlew [hunk 25]
```diff
@@ -161,9 +161,4 @@ function splitJvmOpts() {
 eval splitJvmOpts $DEFAULT_JVM_OPTS $JAVA_OPTS $GRADLE_OPTS
 JVM_OPTS[${#JVM_OPTS[*]}]="-Dorg.gradle.appname=$APP_BASE_NAME"
 
-# by default we should be in the correct project dir, but when run from Finder on Mac, the cwd is wrong
```
Comment (dguy) [inline gradlew:164]: Any reason why these lines have been removed?

File: gradlew.bat [hunk 26]
```diff
@@ -49,6 +49,7 @@ goto fail
 @rem Get command-line arguments, handling Windows variants
 
 if not "%OS%" == "Windows_NT" goto win9xME_args
+if "%@eval[2+2]" == "4" goto 4NT_args
```
Comment (dguy) [inline gradlew.bat:52]: it is not clear to me why this file and `gradlew` are being changed

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultKafkaClientSupplier.java [hunk 27]
```diff
@@ -27,6 +25,8 @@
 import org.apache.kafka.common.serialization.ByteArraySerializer;
 import org.apache.kafka.streams.KafkaClientSupplier;
 
+import java.util.Map;
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultKafkaClientSupplier.java:28]: nit: revert this file back to what it was originally as the only thing that has changed now is the order of the imports

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java [hunk 28]
```diff
@@ -17,259 +17,150 @@
...
-    private static final String ZK_ENTITY_CONFIG_PATH = "/config/topics";
-    // TODO: the following LogConfig dependency should be removed after KIP-4
     public static final String CLEANUP_POLICY_PROP = "cleanup.policy";
     private static final Set<String> CLEANUP_POLICIES = Utils.mkSet("compact", "delete");
     public static final String RETENTION_MS = "retention.ms";
     public static final Long WINDOW_CHANGE_LOG_ADDITIONAL_RETENTION_DEFAULT = TimeUnit.MILLISECONDS.convert(1, TimeUnit.DAYS);
 
-    private final ZkClient zkClient;
-    private final int replicationFactor;
-    private final long windowChangeLogAdditionalRetention;
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:61]: Bad merge? We need this field

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java [hunk 29]
```diff
@@ -17,259 +17,150 @@
...
+        this.replicationFactor = 0;
     }
 
-    public InternalTopicManager() {
-        this.zkClient = null;
+    public InternalTopicManager(StreamsConfig config) {
+        this.config = config;
+        this.streamsKafkaClient = new StreamsKafkaClient(config);
         this.replicationFactor = 0;
-        this.windowChangeLogAdditionalRetention = WINDOW_CHANGE_LOG_ADDITIONAL_RETENTION_DEFAULT;
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:109]: Again, i think you've incorrectly merged. We need this field

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java [hunk 30]
```diff
@@ -17,259 +17,150 @@
...
     }
 
-    public InternalTopicManager(String zkConnect, final int replicationFactor, long windowChangeLogAdditionalRetention) {
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:112]: as above

File: streams/src/test/java/org/apache/kafka/streams/integration/InternalTopicIntegrationTest.java (hunk 15 above)
Comment (enothereska) [reply streams/src/test/java/org/apache/kafka/streams/integration/InternalTopicIntegrationTest.java]: So this remains an issue for me, we shouldn't remove this test.

**Comment by dguy** [comment]
@hjafarpour it looks to me that there has been quite a bad merge. Code is missing that is required for streams to work as expected. In particular `StreamsPartitionAssignor`, `InternalTopicManger` (maybe others). Also, tests have been removed from `InternalTopicIntegrationTest` - these are important tests that prove we have configured the changelogs etc correctly. You need to look at rectifying the merge issue before we can proceed any further. Thanks

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsKafkaClient.java (hunk 20 above)
Comment (hjafarpour) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsKafkaClient.java:66]: We thought we may need a kafka client for other use cases in streams so that was the reason to have a class for it.

Comment (hjafarpour) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsKafkaClient.java:66]: Also I moved the logic for sending request to this class as @dguy  suggested.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java (hunk 28 above)
Comment (hjafarpour) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:61]: Still cleaning it up :)

**Comment by hjafarpour** [comment]
Yes @dguy, I am still cleaning up the merge. Have been busy with other work. Will finish this by tomorrow for your review. Thanks.

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java (hunk 1 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:143]: I think we can just add these configs as part of the PR.

**Comment by dguy** [comment]
ok, ping me when it is ready to be reviewed again. Thanks

**Comment by hjafarpour** [comment]
@dguy Just pushed the changes. it is ready for your review.

**Comment by dguy** [comment]
@hjafarpour i'm confused - this looks like it has a-lot of changes that are unrelated to the task. 266 files changed and +10,326 lines.

**Comment by enothereska** [comment]
Looks like the merge has gone bad, would it be worth doing in another branch?

**Comment by hjafarpour** [comment]
@dguy @enothereska Yes, there were many files from other commits and as Eno said it's merge problem. I'll create a new branch and add the changes there as Eno suggested. This should fix the merge issue. Will ping you when it is ready.

**Comment by hjafarpour** [comment]
This was replaced with "KAFKA-4060: Remove zk client dependency in kafka streams #1884"
